commons/src/main/java/org/frankframework/util/CleanerProvider.java

```diff
@@ -0,0 +1,216 @@
+	private static final Cleaner CLEANER = Cleaner.create();
+
+	static {
+		 // Shutdown hook to generate report of leaks at end of process
+		Runtime.getRuntime().addShutdownHook(new Thread(CleanerProvider::logLeakStatistics));
+	}
+
+	private static void logLeakStatistics() {
+		// Force system to run garbage collection to clean up as much as possible and collect as much leak-info before logging it
+		System.gc();
+		try {
+			// Give time to GC thread.
+			Thread.sleep(1000L);
+		} catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+		}
+		// Write a log message with each stack-trace, for finding the traces where problems originate. Order by largest nr of leaks from origin, ascending.
+		LEAK_MAP.entrySet().stream()
+				.sorted(Comparator.comparingInt(entry -> entry.getValue().get()))
+				.forEachOrdered(entry -> LEAK_LOG.warn("Class [{}] has {} leaks recorded from instances created at:", entry.getKey().owningClassName, entry.getValue().get(), entry.getKey()));
+
+		// Count total leaks and write a log message with each stack-trace, for finding the traces where problems originate.
+		int totalLeaks = LEAK_MAP.values().stream()
+				.map(AtomicInteger::get)
+				.reduce(0, Integer::sum);
+		LEAK_LOG.warn("Total of {} leaks from {} locations", totalLeaks, LEAK_MAP.size());
+	}
+
+	/**
+	 * Method to register a cleaning-action for a method.
+	 *
+	 * @param obj Object for which the cleaning-action should be registered
+	 * @param cleaningAction Cleaning Action to register
+	 * @return ID of the cleaning action, which can be used to execute it.
+	 */
+	public static int register(Object obj, Runnable cleaningAction) {
+		CleaningActionWrapper wrapper = new CleaningActionWrapper(obj, cleaningAction);
+		CLEANER_MAP.put(wrapper.actionId, wrapper);
+		wrapper.cleanable = CLEANER.register(obj, wrapper);
+		return wrapper.actionId;
+	}
+
+	/**
+	 * Execute the cleaning action, if it needs to be manually invoked before the object
+	 * is out of scope, for instance from a close-method.
+	 * This de-registers the cleaning action and marks it so that it is not counted as leaked.
+	 *
+	 * @param cleaningAction Cleaning Action to execute.
+	 */
+	public static void clean(Runnable cleaningAction) {
+		if (cleaningAction == null) {
+			return;
+		}
+		int actionId = getActionId(cleaningAction);
+		CleaningActionWrapper wrapper = CLEANER_MAP.get(actionId);
+		if (wrapper != null) {
+			wrapper.cleaned = true;
+			wrapper.cleanable.clean();
+		} else {
+			cleaningAction.run();
+		}
+	}
+
+	/**
+	 * Execute cleaning action by ID.
+	 *
+	 * @param actionId ID of the cleaning action.
+	 */
+	public static void clean(int actionId) {
+		CleaningActionWrapper wrapper = CLEANER_MAP.get(actionId);
+		if (wrapper != null) {
+			wrapper.cleaned = true;
+			wrapper.cleanable.clean();
+		}
+	}
+
+	private static int recordLeak(LeakedResourceException leakedResourceException) {
```

Comment:
Sonarlint zegt: 'Move this method into "CleaningActionWrapper".' -- 'When a private method is only invoked by an inner class, there’s no reason not to move it into that class. It will still have the same access to the outer class' members, but the outer class will be clearer and less cluttered.'